fuse_loss/include/fuse_loss/loss_function.h

```diff
@@ -0,0 +1,284 @@
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef FUSE_LOSS_LOSS_FUNCTION_H
+#define FUSE_LOSS_LOSS_FUNCTION_H
+
+#include <ceres/loss_function.h>
+
+// This provides additional loss functions that are not available in:
+// https://github.com/ceres-solver/ceres-solver/blob/master/include/ceres/loss_function.h
+//
+// Most of them are available in other libraries like:
+// * g2o: https://github.com/RainerKuemmerle/g2o/blob/master/g2o/core/robust_kernel_impl.h
+// * GTSAM: https://github.com/borglab/gtsam/blob/develop/gtsam/linear/LossFunctions.h
+//          (formerly in https://github.com/borglab/gtsam/blob/develop/gtsam/linear/NoiseModel.h)
+//
+// Remember that the loss functions in Ceres solver implement a slight variation of the \rho(r) function, defined for
+// multiple M-estimators in table #1 from http://www.audentia-gestion.fr/research.microsoft/ZhangIVC-97-01.pdf (p. 24).
+//
+// In Ceres the loss functions evaluate:
+//
+//   rho(s) = 2 * \rho(sqrt(s))
+//
+// where s = r^2, being r the residual.
+//
+// This is because the output of rho(s) is mutiplied by 0.5 when the residual block cost is computed, either with or
+// without loss:
+// * cost w/o loss: https://github.com/ceres-solver/ceres-solver/blob/master/internal/ceres/residual_block.cc#L159
+// * cost w/  loss: https://github.com/ceres-solver/ceres-solver/blob/master/internal/ceres/residual_block.cc#L165
+//
+// Note that according to this, it looks like the Tukey loss function is incorrectly implemented in Ceres because it
+// must be multiplied by 2, so instead of dividing by 6 it should divide by 3. See:
```

Comment:
One final question:
Do you want to make a fuse version of the Tukey loss function with the proper scale? Basically a copy&paste from the Ceres implementation, but then divide by the correct value. We could add a TODO note to remove the fuse version if/when the Ceres PR gets merged. That would avoid having to use the Scale+Tukey combo.

Reply:
I'll prefer to wait a little more to get an answer from https://ceres-solver-review.googlesource.com/c/ceres-solver/+/16700 before I do that, so I'd prefer this is merged as is now. I'm happy with the `Scale(2) + Tukey` combo workaround for the time being.

Basically, I want to get the confirmation that their implementation must be fixed.

After that I'd send another PR here anyway, because the fixed version would live only in the latest Ceres version. So I'd send a PR that conditionally implements the Tukey loss function here, or uses Ceres one, based on the Ceres version, as I already did for other stuff.